lib/cylc/task_remote_mgr.py

```diff
@@ -0,0 +1,350 @@
+        timeout = time() + 10.0
+        while procs and time() < timeout:
+            for (host, owner), (cmd, proc) in procs.copy().items():
+                if proc.poll() is None:
+                    continue
+                del procs[(host, owner)]
+                out, err = proc.communicate()
+                if proc.wait():
+                    LOG.warning(TaskRemoteMgmtError(
+                        TaskRemoteMgmtError.MSG_TIDY,
+                        (host, owner), ' '.join(quote(item) for item in cmd),
+                        proc.ret_code, out, err))
+        # Terminate any remaining commands
+        for (host, owner), (cmd, proc) in procs.items():
+            try:
+                proc.terminate()
+            except OSError:
+                pass
+            out, err = proc.communicate()
+            if proc.wait():
+                LOG.warning(TaskRemoteMgmtError(
+                    TaskRemoteMgmtError.MSG_TIDY,
+                    (host, owner), ' '.join(quote(item) for item in cmd),
+                    proc.ret_code, out, err))
+
+    def _remote_host_select_callback(self, proc_ctx, cmd_str):
+        """Callback when host select command exits"""
+        self.ready = True
+        if proc_ctx.ret_code == 0 and proc_ctx.out:
+            # Good status
+            LOG.debug(proc_ctx)
+            self.remote_host_str_map[cmd_str] = proc_ctx.out.splitlines()[0]
+        else:
+            # Bad status
+            LOG.error(proc_ctx)
+            self.remote_host_str_map[cmd_str] = TaskRemoteMgmtError(
+                TaskRemoteMgmtError.MSG_SELECT, (cmd_str, None), cmd_str,
+                proc_ctx.ret_code, proc_ctx.out, proc_ctx.err)
+
+    def _remote_init_callback(self, proc_ctx, host, owner, tmphandle):
+        """Callback when "cylc remote-init" exits"""
+        self.ready = True
+        tmphandle.close()
+        if proc_ctx.ret_code == 0:
+            for status in (REMOTE_INIT_DONE, REMOTE_INIT_NOT_REQUIRED):
+                if status in proc_ctx.out:
+                    # Good status
+                    LOG.debug(proc_ctx)
+                    self.remote_init_map[(host, owner)] = status
+                    return
+        # Bad status
+        LOG.error(TaskRemoteMgmtError(
+            TaskRemoteMgmtError.MSG_INIT,
+            (host, owner), ' '.join(quote(item) for item in proc_ctx.cmd),
+            proc_ctx.ret_code, proc_ctx.out, proc_ctx.err))
+        LOG.error(proc_ctx)
+        self.remote_init_map[(host, owner)] = REMOTE_INIT_FAILED
+
+    def _remote_init_items(self, host, owner):
+        """Return list of items that should be installed on task remote.
+
+        Each item is (path, name),
+        where name is relative path under suite run directory.
+        """
+        items = []
+        comm_meth = GLOBAL_CFG.get_host_item(
+            'task communication method', host, owner)
+        LOG.debug('comm_meth=%s' % comm_meth)
```

Comment:
Is this needed any more?

Reply:
OK for debug purpose. I am sure we'll revisit this when we tackle #2214.